File: backend/app/services/selector_ranking.py
"""候选结构、身份与排序契约；镜像 frontend/src/utils/selector-ranking.js。
结构合法不等于现场唯一。name/exact 属于定位身份，不能在回填或去重时丢弃。
"""
from __future__ import annotations
import json
import re

VALID_BYS = {"testid", "xpath", "role", "label", "text", "placeholder", "css"}
FRAGILE_BYS = {"text", "role"}


def is_valid_candidate(cand: dict) -> bool:
    return (isinstance(cand, dict) and cand.get("by") in VALID_BYS
            and isinstance(cand.get("value"), str) and bool(cand["value"].strip())
            and ("name" not in cand or isinstance(cand["name"], str))
            and ("exact" not in cand or type(cand["exact"]) is bool))


def normalize_candidate(cand: dict) -> dict | None:
    if not is_valid_candidate(cand):
        return None
    return {k: cand[k] for k in ("by", "value", "name", "exact", "src", "status", "disabled") if k in cand}


def candidate_identity(cand: dict) -> str:
    # 历史探测用 CSS 写 data-testid，录制用 testid；两种精确写法是同一定位。
    if cand.get("by") == "css":
        match = re.fullmatch(r'\[data-testid=("(?:[^"\\]|\\.)*"|[\w-]+)\]', cand.get("value", ""))
        if match:
            value = match[1]
            try:
                cand = {**cand, "by": "testid", "value": json.loads(value) if value.startswith('"') else value}
            except ValueError:
                pass
    return json.dumps([cand.get("by"), cand.get("value"), cand.get("name") or None,
                       cand.get("exact", False)], ensure_ascii=False, separators=(",", ":"))


def is_active_candidate(cand: dict) -> bool:
    return (is_valid_candidate(cand) and cand.get("src") != "learned"
            and cand.get("status") not in ("pending", "rejected", "retired")
            and cand.get("disabled") is not True)


def valid_candidates(cands: list[dict]) -> list[dict]:
    return [c for c in cands if is_valid_candidate(c)] if isinstance(cands, list) else []


def is_fragile(cand: dict) -> bool:
    return (cand.get("by") == "role" and not (cand.get("name") and cand.get("exact"))) or (
        cand.get("by") == "text" and not cand.get("exact"))


def candidate_rank(cand: dict) -> int:
    by, value = cand.get("by"), cand.get("value", "")
    if by == "testid": return 0
    if by == "role" and cand.get("name") and cand.get("exact"): return 1
    if by == "label": return 2
    if by == "placeholder": return 3
    if by == "css" and (value.startswith("#") or value.startswith("[data-test")): return 4
    if by == "text" and cand.get("exact"): return 5
    if by == "role" and cand.get("name"): return 6
    if by == "xpath": return 7
    if by == "css": return 8
    return 9


def order_candidates(cands: list[dict]) -> list[dict]:
    return sorted(cands, key=candidate_rank)


def merge_candidates(*groups: list[dict], limit: int = 6) -> list[dict]:
    seen, merged = set(), []
    for group in groups:
        for c in valid_candidates(group):
            identity = candidate_identity(c)
            if identity not in seen:
                seen.add(identity)
                merged.append(normalize_candidate(c))
    return order_candidates(merged)[:limit]

File: frontend/src/utils/selector-ranking.js
// 镜像后端 selector_ranking.py；现场唯一性由探测/runtime 验证。
export const FRAGILE_BYS = new Set(['text', 'role'])
const VALID_BYS = new Set(['testid', 'xpath', 'role', 'label', 'text', 'placeholder', 'css'])
export function normalizeCandidate(c) {
  if (!c || !VALID_BYS.has(c.by) || typeof c.value !== 'string' || !c.value.trim()
      || ('name' in c && typeof c.name !== 'string') || ('exact' in c && typeof c.exact !== 'boolean')) return null
  return Object.fromEntries(['by', 'value', 'name', 'exact', 'src', 'status', 'disabled']
    .filter(k => k in c).map(k => [k, c[k]]))
}
export const candidateIdentity = c => {
  const match = c.by === 'css' && c.value?.match(/^\[data-testid=("(?:[^"\\]|\\.)*"|[\w-]+)\]$/)
  if (match) { try { c = { ...c, by: 'testid', value: match[1].startsWith('"') ? JSON.parse(match[1]) : match[1] } } catch {} }
  return JSON.stringify([c.by, c.value, c.name || null, c.exact ?? false])
}
export const isActiveCandidate = c => !!normalizeCandidate(c) && c.src !== 'learned'
  && !['pending', 'rejected', 'retired'].includes(c.status) && c.disabled !== true
export function isFragile(c) {
  return (c?.by === 'role' && !(c.name && c.exact)) || (c?.by === 'text' && !c.exact)
}
export function candidateRank(c) {
  if (c.by === 'testid') return 0
  if (c.by === 'role' && c.name && c.exact) return 1
  if (c.by === 'label') return 2
  if (c.by === 'placeholder') return 3
  if (c.by === 'css' && (c.value.startsWith('#') || c.value.startsWith('[data-test'))) return 4
  if (c.by === 'text' && c.exact) return 5
  if (c.by === 'role' && c.name) return 6
  if (c.by === 'xpath') return 7
  if (c.by === 'css') return 8
  return 9
}
export const orderCandidates = cands => [...(cands || [])].sort((a, b) => candidateRank(a) - candidateRank(b))
export function mergeCandidates(...groups) {
  const unique = new Map()
  for (const c of groups.flat().map(normalizeCandidate).filter(Boolean)) {
    const id = candidateIdentity(c)
    if (!unique.has(id)) unique.set(id, c)
  }
  return orderCandidates([...unique.values()]).slice(0, 6)
}
